File: concept/Инкапсуляция/examples/python/пример.py
class BankAccount:
    """
    Класс банковского счёта.
    Демонстрирует принцип инкапсуляции.
    """

    def __init__(self, owner: str, balance: float):
        self.owner = owner
        self.__balance = balance

    @property
    def balance(self) -> float:
        """
        Геттер для получения баланса.
        """
        return self.__balance

    def deposit(self, amount: float) -> None:
        """
        Пополнение счёта.
        """
        if amount > 0:
            self.__balance += amount
            print(f"Счёт пополнен на {amount}")
        else:
            print("Сумма должна быть положительной")

    def withdraw(self, amount: float) -> None:
        """
        Снятие средств со счёта.
        """
        if 0 < amount <= self.__balance:
            self.__balance -= amount
            print(f"Снято {amount}")
        else:
            print("Недостаточно средств")


account = BankAccount("Алексей", 1000)

print(f"Текущий баланс: {account.balance}")

account.deposit(500)
print(f"Баланс после пополнения: {account.balance}")

account.withdraw(300)
print(f"Баланс после снятия: {account.balance}")

# Прямой доступ запрещён
# print(account.__balance)  # Ошибка

File: concept/Инкапсуляция/examples/C/пример.c
#include <stdio.h>
#include <string.h>

/*
 * Демонстрация инкапсуляции на языке C.
 * В C нет встроенной поддержки ООП,
 * поэтому инкапсуляция реализуется через структуры и функции.
 */

typedef struct {
    char owner[50];
    double balance;
} BankAccount;


/*
 * Создание банковского счёта
 */
void init_account(BankAccount *account, const char *owner, double balance) {
    strcpy(account->owner, owner);
    account->balance = balance;
}


/*
 * Получение баланса
 */
double get_balance(BankAccount *account) {
    return account->balance;
}


/*
 * Пополнение счёта
 */
void deposit(BankAccount *account, double amount) {
    if (amount > 0) {
        account->balance += amount;
        printf("Счёт пополнен на %.2f\n", amount);
    } else {
        printf("Сумма должна быть положительной\n");
    }
}


/*
 * Снятие средств
 */
void withdraw(BankAccount *account, double amount) {
    if (amount > 0 && amount <= account->balance) {
        account->balance -= amount;
        printf("Снято %.2f\n", amount);
    } else {
        printf("Недостаточно средств\n");
    }
}


int main() {

    BankAccount account;

    init_account(&account, "Алексей", 1000);

    printf("Текущий баланс: %.2f\n", get_balance(&account));

    deposit(&account, 500);
    printf("Баланс после пополнения: %.2f\n", get_balance(&account));

    withdraw(&account, 300);
    printf("Баланс после снятия: %.2f\n", get_balance(&account));

    return 0;
}
